src/linear_layer.cpp

```diff
@@ -635,3 +638,241 @@ std::unique_ptr<BaseLayer> Linear::to_cuda() {
+                C_zo_zo += cov_hh[m] * mu_w[t] * mu_w[j];
+            } else {
+                C_zo_zo +=
+                    var_w[t] * (cov_hh[m] + mu_h_prev[t] * mu_h_prior[j]) +
+                    cov_hh[m] * mu_w[t] * mu_w[j];
+            }
+        }
+    }
+    C_zo_zo = C_zo_zo + var_b[0];
+    cov_zo.push_back(C_zo_zo);
+}
+
+void slinear_smoother_z_ouput(int num_timestep, std::vector<float> &cov,
+                              std::vector<float> &mu_priors,
+                              std::vector<float> &var_priors,
+                              std::vector<float> &mu_posts,
+                              std::vector<float> &var_posts,
+                              std::vector<float> &mu_smooths,
+                              std::vector<float> &var_smooths)
+/*
+ */
+{
+    for (int i = num_timestep - 2; i >= 0; i--) {
+        float tmp = cov[i + 1] / var_priors[i + 1];
+        mu_smooths[i] =
+            mu_posts[i] + tmp * (mu_smooths[i + 1] - mu_priors[i + 1]);
+        var_smooths[i] =
+            var_posts[i] + tmp * (var_smooths[i + 1] - var_priors[i + 1]) * tmp;
+    }
+}
+
+void SLinear::forward(BaseHiddenStates &input_states,
+                      BaseHiddenStates &output_states,
+                      BaseTempStates &temp_states)
+/*
+ */
+{
+    // Initialization
+    int batch_size = input_states.block_size;
+    this->set_cap_factor_udapte(batch_size);
+    if (temp_states.slinear.num_hs == 0) {
+        temp_states.slinear.set_num_states(this->input_size);
+    }
+
+    // Forward pass
+    if (this->num_threads > 1) {
+        linear_fwd_mean_var_mp(this->mu_w, this->var_w, this->mu_b, this->var_b,
+                               input_states.mu_a, input_states.var_a,
+                               this->input_size, this->output_size, batch_size,
+                               this->bias, this->num_threads,
+                               output_states.mu_a, output_states.var_a);
+    } else {
+        int start_chunk = 0;
+        int end_chunk = this->output_size * batch_size;
+        linear_fwd_mean_var(this->mu_w, this->var_w, this->mu_b, this->var_b,
+                            input_states.mu_a, input_states.var_a, start_chunk,
+                            end_chunk, this->input_size, this->output_size,
+                            batch_size, this->bias, output_states.mu_a,
+                            output_states.var_a);
+    }
+    // Update number of actual states.
+    output_states.width = this->out_width;
+    output_states.height = this->out_height;
+    output_states.depth = this->out_channels;
+    output_states.block_size = batch_size;
+    output_states.actual_size = this->output_size;
+
+    // save z_output prior for smoothing
+    slinear_states.mu_zo_priors.push_back(output_states.mu_a[0]);
+    slinear_states.var_zo_priors.push_back(output_states.var_a[0]);
+
+    // save cov_zo for smoother
+    slinear_cov_zo_smoother(this->input_size, this->output_size, this->mu_w,
```

Comment:
We shouldn't store any vector that will later be used in `backward()` or `smoother()` because it is temporary state. It is there for holding temporary values in a short term and it will be erased when it gets called. In this case, you should declare all the vectors needed for smoother step in `slinear_states`

Reply:
I was not clear in the function name. These vectors are saved in `SLSTM` forward and later used in `SLinear` forward pass only. Not in `backward()` and `smoother()`. See above comment, it all boils down to modify `BaseTempStates` or `BaseHiddenStates`.

Reply:
See comment above